src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -390,8 +497,145 @@ fn compute_deps_custom_build(
         // Build scripts always compiled for the host.
         CompileKind::Host,
         CompileMode::Build,
+        IS_NO_ARTIFACT_DEP,
     )?;
-    Ok(vec![unit_dep])
+
+    let artifact_build_deps = state.deps(unit, script_unit_for, &|_unit, dep| {
+        dep.kind() == DepKind::Build && dep.artifact().is_some()
+    });
```

Comment:
That's a great idea, I like that much more as well.

A first step towards implementing this is in 61b285cdd4a2ff45a5d830b3a95f6496f9fa4601 and c0e6abe384c2c6282bdd631e2f2a3b092043e6c6, and more will follow to cleanup up filter handling.

Reply:
Maybe in conjunction with the [my reply above](https://github.com/rust-lang/cargo/pull/9992#discussion_r808869032), while working on this I thought maybe it's possible to push the entire handling of artifact deps 'down' into `compute_deps()` with the special casing needed to deal with the targets. This would mean that no special `deps_filtered()` would be required anymore. Having looked into this too it seems that the reason this code is there in the first place is that maybe it's hard to unambiguously detect the case.

If you think it's desirable enough I could give it another shot.